### PR TITLE
Exec loop minor performance improvement

### PR DESCRIPTION
**Pull request by mateusz-sekara**
Moving all the token misc calls that need to be done before the exec loop to a separate function should reduce the number of details in the exec loop. 
What is more, we can call that at the beginning of processing - this should speed up filling the cache with data (snoozed roots). The previous implementation kept fetching all these token details even though it might be terminated immediately because all messages were already executed.

### CODE REVIEW

File: core/services/ocr2/plugins/ccip/ccipexec/ocr2.go [hunk 1]
```diff
@@ -266,11 +211,11 @@ func (r *ExecutionReportingPlugin) getExecutableObservations(ctx context.Context
...
 				rep,
 				inflight,
-				allowedTokenAmountValue.Tokens,
-				sourceTokensPricesValue,
-				destTokensPricesValue,
-				getDestGasPrice,
-				sourceToDestTokens,
+				token.rateLimiterTokenBucket.Tokens,
+				token.sourceTokenPrices,
+				token.destTokenPrices,
```
Comment (dimkouv) [inline core/services/ocr2/plugins/ccip/ccipexec/ocr2.go]: I suggest renaming `token`.
`token.destTokenPrices` seems weird.

File: core/services/ocr2/plugins/ccip/ccipexec/ocr2.go [hunk 2]
```diff
@@ -1079,6 +1017,82 @@ func (r *ExecutionReportingPlugin) getUnexpiredCommitReports(
...
+	sourceTokenPrices      map[common.Address]*big.Int
+	destTokenPrices        map[common.Address]*big.Int
+	sourceToDestTokens     map[common.Address]common.Address
+	gasPrice               prices.GasPrice
+}
+
+// prepareTokenExecData gather all the pre-execution data needed for token execution into a single lazy call.
+// This is done to avoid fetching the data multiple times for each message. Additionally, most of the RPC calls
+// within that function is cached, so it should be relatively fast and not require any RPC batching.
+func (r *ExecutionReportingPlugin) prepareTokenExecData(ctx context.Context) cache.LazyFunction[execTokenData] {
```
Comment (dimkouv) [inline core/services/ocr2/plugins/ccip/ccipexec/ocr2.go]: Based on the method name, i would expect that `prepareTokenExecData` does something. But in fact it returns a function that does something.

I suggest either returning `execTokenData` and wrap that in a LazyFunction on the caller. Or renaming it.

File: core/services/ocr2/plugins/ccip/ccipexec/ocr2.go [hunk 3]
```diff
@@ -137,66 +137,25 @@ func (r *ExecutionReportingPlugin) getExecutableObservations(ctx context.Context
...
+		if err1 != nil {
+			return nil, err1
 		}
 
 		getDestPoolRateLimits := cache.LazyFetch(func() (map[common.Address]*big.Int, error) {
-			sourceToDestTokens, err := getSourceToDestTokens()
-			if err != nil {
-				return nil, err
+			tokens, err2 := getExecTokenData()
+			if err2 != nil {
```
Comment (dimkouv) [inline core/services/ocr2/plugins/ccip/ccipexec/ocr2.go]: I don't like having `err2` everywhere, I believe we can use `err` with a different workaround for the linter.

**Review by dimkouv — APPROVED** [review]
logic lgtm, i suggest fixing the nits i raised

File: core/services/ocr2/plugins/ccip/ccipexec/ocr2.go (hunk 3 above)
Comment (mateusz-sekara) [reply core/services/ocr2/plugins/ccip/ccipexec/ocr2.go]: What do you suggest instead?

**Pull request merged**